master/internal/provisioner/scale_decider.go

```diff
@@ -110,12 +127,37 @@ func (s *scaleDecider) findInstancesToTerminate(
 			toTerminate[inst.ID] = true
 
 		case Running:
-			if _, ok := s.agentSnapshot[inst.AgentName]; ok {
+			if _, ok := s.idleAgentSnapshot[inst.AgentName]; ok {
 				idleInstances[inst.ID] = true
 			}
 		}
 	}
 
+	// Terminate instances that have not connected to the master beyond the max startup period.
+	now := time.Now()
+	for _, inst := range s.instanceSnapshot {
+		if s.launchTime.Add(masterLaunchConnectionBuffer).After(now) {
```

Comment:
blocking: It's confusing to check if the provisioner recently launches some instances we ignore the unconnected agents.

Reply:
Actually this completely went away in the new version.